server/channels/jobs/s3_path_migration/s3_path_migration.go

```diff
@@ -0,0 +1,276 @@
+	// If the type cast fails, it will be nil
+	// which is checked later.
+	s3Backend, _ := fileBackend.(*filestore.S3FileBackend)
+	worker := &S3PathMigrationWorker{
+		jobServer:   jobServer,
+		store:       store,
+		fileBackend: s3Backend,
+		name:        JobName,
+		stop:        make(chan bool, 1),
+		stopped:     make(chan bool, 1),
+		jobs:        make(chan model.Job),
+	}
+	return worker
+}
+
+func (worker *S3PathMigrationWorker) Run() {
+	mlog.Debug("Worker started", mlog.String("worker", worker.name))
+
+	defer func() {
+		mlog.Debug("Worker finished", mlog.String("worker", worker.name))
+		worker.stopped <- true
+	}()
+
+	for {
+		select {
+		case <-worker.stop:
+			mlog.Debug("Worker received stop signal", mlog.String("worker", worker.name))
+			return
+		case job := <-worker.jobs:
+			mlog.Debug("Worker received a new candidate job.", mlog.String("worker", worker.name))
+			worker.DoJob(&job)
+		}
+	}
+}
+
+func (worker *S3PathMigrationWorker) Stop() {
+	mlog.Debug("Worker stopping", mlog.String("worker", worker.name))
+	close(worker.stop)
+	<-worker.stopped
+}
+
+func (worker *S3PathMigrationWorker) JobChannel() chan<- model.Job {
+	return worker.jobs
+}
+
+func (worker *S3PathMigrationWorker) IsEnabled(_ *model.Config) bool {
+	return os.Getenv("MM_CLOUD_FILESTORE_BIFROST") != ""
+}
+
+func (worker *S3PathMigrationWorker) getJobMetadata(job *model.Job, key string) (int, *model.AppError) {
+	countStr := job.Data[key]
+	count := 0
+	var err error
+	if countStr != "" {
+		count, err = strconv.Atoi(countStr)
+		if err != nil {
+			return 0, model.NewAppError("getJobMetadata", model.NoTranslation, nil, "", http.StatusInternalServerError).Wrap(err)
+		}
+	}
+	return count, nil
+}
+
+func (worker *S3PathMigrationWorker) DoJob(job *model.Job) {
+	defer worker.jobServer.HandleJobPanic(job)
+
+	if claimed, err := worker.jobServer.ClaimJob(job); err != nil {
+		mlog.Warn("S3PathMigrationWorker experienced an error while trying to claim job",
+			mlog.String("worker", worker.name),
+			mlog.String("job_id", job.Id),
+			mlog.Err(err))
+		return
+	} else if !claimed {
+		return
+	}
+
+	if worker.fileBackend == nil {
+		mlog.Error("S3PathMigrationWorker: no s3 file backend found.")
+		return
+	}
+
+	var appErr *model.AppError
+	// We get the job again because ClaimJob changes the job status.
+	job, appErr = worker.jobServer.GetJob(job.Id)
+	if appErr != nil {
+		mlog.Error("S3PathMigrationWorker: job execution error", mlog.String("worker", worker.name), mlog.String("job_id", job.Id), mlog.Err(appErr))
+		worker.setJobError(job, appErr)
+		return
+	}
+
+	totalCount, appErr := worker.getJobMetadata(job, "total_file_count")
+	if appErr != nil {
+		mlog.Error("S3PathMigrationWorker: failed to get total file count", mlog.String("worker", worker.name), mlog.String("job_id", job.Id), mlog.Err(appErr))
+		worker.setJobError(job, appErr)
+		return
+	}
+	if totalCount == 0 {
+		total, err := worker.store.FileInfo().CountAll()
```

Comment:
I see we use the `totalCount` to measure job progress, but given we're starting only with the past four weeks, this won't be accurate. Is there another denominator we could use to track progress?

Reply:
Lol .. my bad there. You are right. We could obviously count files from the given time, but I doubt that it's worth it, given we are anyways storing the number of files processed in the job metadata. I am of the opinion to just keep progress as 0, and then move to 100 when done unless you have strong objections.

Reply:
No objection to simplifying the progress!